config/package_resouce.go

```diff
@@ -82,164 +92,264 @@ func (p *packageResouce) validate() (*ManagedResources, error) {
-		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
-			p.managedPackage.Zypper.Install = pr
-		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
-			p.managedPackage.Zypper.Remove = pr
-		}
+		p.managedPackage.Zypper = &ZypperPackage{DesiredState: p.GetDesiredState(), PackageResource: pr}
+
 	case *agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_Rpm:
 		pr := p.GetRpm()
 		if !packages.RPMExists {
 			return nil, fmt.Errorf("cannot manage RPM package because rpm does not exist on the system")
 		}
 
-		switch p.GetDesiredState() {
-		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_INSTALLED:
-			p.managedPackage.RPM.Install = pr
-		case agentendpointpb.ApplyConfigTask_Config_Resource_PackageResource_REMOVED:
-			p.managedPackage.RPM.Remove = pr
-		}
+		p.managedPackage.RPM = &RPMPackage{DesiredState: p.GetDesiredState(), PackageResource: pr}
+
 	default:
 		return nil, errors.New("SystemPackage field not set or references unknown package manager")
 	}
 
 	return &ManagedResources{Packages: []ManagedPackage{p.managedPackage}}, nil
 }
 
-// TODO: implement a caching system for installed packages.
-var aptInstalled []packages.PkgInfo
-var debInstalled []packages.PkgInfo
-var gooInstalled []packages.PkgInfo
-var yumInstalled []packages.PkgInfo
-var zypperInstalled []packages.PkgInfo
-var rpmInstalled []packages.PkgInfo
-
-func pkgInstalled(installedPkgs []packages.PkgInfo, name string) bool {
-	for _, pkg := range installedPkgs {
-		if pkg.Name == name {
-			return true
-		}
+var aptInstalled map[string]struct{}
+var debInstalled map[string]struct{}
+var gooInstalled map[string]struct{}
+var yumInstalled map[string]struct{}
+var zypperInstalled map[string]struct{}
+var rpmInstalled map[string]struct{}
+
+func populateInstalledCache(ctx context.Context, mp ManagedPackage) error {
+	var cache *map[string]struct{}
+	var refreshFunc func(context.Context) ([]packages.PkgInfo, error)
+	var err error
+	switch {
+	// TODO: implement apt functions
+	case mp.Apt != nil:
+		cache = &aptInstalled
+		refreshFunc = packages.InstalledDebPackages
+
+	case mp.Deb != nil:
+		cache = &debInstalled
+		refreshFunc = packages.InstalledDebPackages
+
+	case mp.GooGet != nil:
+		cache = &gooInstalled
+		refreshFunc = packages.InstalledGooGetPackages
+
+	// TODO: implement msi functions
+	case mp.MSI != nil:
+		return errors.New("msi not implemented")
+
+	// TODO: implement yum functions
+	case mp.Yum != nil:
+		cache = &yumInstalled
+		refreshFunc = packages.InstalledRPMPackages
+
+	// TODO: implement zypper functions
+	case mp.Zypper != nil:
+		cache = &zypperInstalled
+		refreshFunc = packages.InstalledRPMPackages
+
+	case mp.RPM != nil:
+		cache = &rpmInstalled
+		refreshFunc = packages.InstalledRPMPackages
+	default:
+		return fmt.Errorf("unknown or unpopulated ManagedPackage package type: %+v", mp)
+	}
+
+	// Cache already populated.
+	if *cache != nil {
+		return nil
+	}
```

Comment:
When the cache is already populated, when do we re-populate it?

If a package is installed outside the config management system, would the cache reflect that?

Reply:
I was going back and forth on how to implement this, for now i created a catch type with a refreshed timestamp